### PR TITLE
Add length and padding keyworks to DistributedSampler

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Current implementation of `DistributedSampler` is ideal for distributed
training using map datasets, as they fit in memory and have known size.
However, it doesn't support distributed training using `IterableDataset`
datasets, as these classes do not implement `__len__`.
To fix that, a `length` keyword was added to `DistributedSampler`, which
has precedence when set.

An extra `padding=True` parameter was also added was give finer control
on whether the (returned) index list should be padded by the sampler.
This is useful for preventing duplicate reading on `IterableDataset`
datasets that do not fit in memory or which data reading or transformation
are expensive.

Finally, `set_rank` method was added to ease distributed training, allowing `DataLoader`'s worker processes to register themselves on `DistributedSampler` instances through `worker_init_fn` method.  This is useful when worker processes want to change the sampling behavior based on not only in the process rank, but also on their worker ID. [IterableDataset documentation](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataset.py#L44-L52) mentions this issue, but the examples on it only handle dataset with known size, which is not always the case for `IterableDataset` dataset.

There is no back compatibility with this change.

### CODE REVIEW

File: torch/utils/data/distributed.py
Comment (vincentqb) [inline torch/utils/data/distributed.py]: I read that `num_samples` is the length of the padded dataset, no? If so, I would expect to get `self._dataset_length` here then. Would this break BC?

**Review by vincentqb — CHANGES_REQUESTED** [review]

**Comment by vincentqb** [comment]
CC @cpuhrsch for #26547.

File: torch/utils/data/distributed.py
Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: Thanks for the quick reply! Good catch. Current implementation was always returning padded length. I have tweaked it so it returns `self._dataset_length` when `padding=False`. This doesn't break BC because `padding=True` by default.

**Comment by thiagocrepaldi** [comment]
Out of curiosity, are the `*.pyi` files automatically generated? If so, which tool is used?

**Comment by vincentqb** [comment]
> Out of curiosity, are the `*.pyi` files automatically generated? If so, which tool is used?

Manually :)

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [inline torch/utils/data/distributed.py]: Could you add a bit more info on how this "same length" is achieved?

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [inline torch/utils/data/distributed.py]: I think it's better to throw an error here. We want to explicitly check that, if length is not None, it is indeed an integer and indeed greater than 0, otherwise throw an error.

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [inline torch/utils/data/distributed.py]: This seems strange. Why isn't this also num samples?

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [inline torch/utils/data/distributed.py]: This seems like a bigger change to how the sampler is used. Outside of your intended use-case, how is this useful and why is this necessary?

**Comment by cpuhrsch** [comment]
Thanks for splitting this out!

File: torch/utils/data/distributed.py
Comment (vincentqb) [reply torch/utils/data/distributed.py]: Indeed. Should `_dataset_length` just be replaced by `num_samples`?

Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: `DistributedSample`._len__ will typically be used by map style datasets only. given an index list such as `[1,2,3,4,5,6,7]` and `num_replicas=3` as input, sampler should return `[[1,2,3],[4,5,6],[7,1,2]]` and `len(sampler)=3` for every batch

When `padding=False`, on the other hand, given the same output, the sampler must return `[1,2,3],[4,5,6],[7]` but its length varies and for `IterableDataset` not really used.

Returning `self.num_samples` works for me. I can add a note informing that regardless of `padding` value, :meth:`__len__` will return `ceil(dataset length / num_replicas)`

File: torch/utils/data/distributed.py
Comment (apaszke) [inline torch/utils/data/distributed.py]: Doing `DistributedSampler(None, length=5)` is weird, why can't we allow an `int` in place of the dataset?

File: torch/utils/data/distributed.py
Comment (apaszke) [inline torch/utils/data/distributed.py]: Please don't use RST-style references (`:attr:`) in error messages!

File: torch/utils/data/distributed.py
Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: After consolidating all feedbacks to eliminate `DistributedChunkSampler` from https://github.com/pytorch/pytorch/pull/26547 and after doing local tests, I noticed that we need this setter along with the other changes in this PR.

In a nutshell, for distributed training using `ChunkDataset` datasets with multiple dataloader workers, every process will have a copy of the dataset, which has a copy of the `DistributedSampler`. The initial value of sampler's `rank` must be `None` as it is the `DataLoader` who knows how many `num_workers` there will be. So we need a setter so that each process can assign their rank ID to the sampler (https://github.com/pytorch/pytorch/blob/8362a13a4e9791923ac5967506dc41a9a24e5be7/torch/utils/data/dataset.py#L373 and https://github.com/pytorch/pytorch/blob/8362a13a4e9791923ac5967506dc41a9a24e5be7/torch/utils/data/dataset.py#L327 shows their use)

File: torch/utils/data/distributed.py
Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: Both approaches work for our use case. I can make the change in a BC way if `dataset` being a `int` and not an instance of `Dataset` feels clearer. If more context is needed, this design was initially discussed [here](https://github.com/pytorch/pytorch/pull/26547#issuecomment-546480531)

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [reply torch/utils/data/distributed.py]: Sounds good to me! - Did a missclick on the frowny face....

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [reply torch/utils/data/distributed.py]: @apaszke - aside from the reasoning in my comment, branching on the type of an input can also make it easier to make a mistake. Even if we added the option of allowing an int in place of the dataset, it should still make sense to provide the length argument as an additional feature for existing API calls (i.e. those that pass in a map-style dataset).

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [reply torch/utils/data/distributed.py]: You're passing in the sampler during construction of the DataLoader I assume (since DataLoader takes a sampler argument), so you will know the number of workers. Why not use that information?

Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: We cannot pass any sampler to the `DataLoader` when `IterableDataset` is used - by design. That is actually the reason why we need an external sampler and the ability to set rank after DataLoader spawns workers

Yes, `set_rank` allows workers to register themselves after DataLoader creates them. In a similar fashion, the existing `set_epoch` method allows the training loop to register the current epoch with the sampler too.

**Comment by thiagocrepaldi** [comment]
@cpuhrsch Did you have a chance to check my last replies regarding using `dataset` for both `int`/`Dataset` types (or adding a new `length` keyword) and why `set_rank` is needed?

File: torch/utils/data/distributed.py
Comment (cpuhrsch) [reply torch/utils/data/distributed.py]: Sounds like we should be able to add a sampler to DataLoader. For example, we might want to suppor reservoir samplers. Let's change the DataLoader to support this or figure out why it can't.

Comment (cpuhrsch) [reply torch/utils/data/distributed.py]: See https://github.com/pytorch/pytorch/issues/28743

Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: `DataLoader`'s sampler was originally meant to return example indices, which works great for map-style datasets with known size.

After `IterableDataset` was introduced, on the other hand, `DataLoader` doesnt  know the size of the iterable dataset and it must iterate over it until `StopIteration` is raised. In this case, it makes sense to disable external samplers and create an internal sampler, but at the same time a gap with sampling was created.

In order to add back `sampler` to `DataLoader`, we would have to make it clear that the sampler will be used by `IterableDataset` in the worker process as opposed to be used by `DataLoader` (as it would be if dataset was map-style).

In both cases (modifying `DistributedSampler` or changing `DataLoader` to accept sampler for IterableDataset), in order to prevent `DataLoader` to be re-instantiated every epoch, destroying all workers context, a `set_rank` would be needed so that worker processes can use it.

Therefore, IMHO, this PR is still valid as `padding`, `length` and `set_rank` would be used by both approaches: `Dataloader.sampler` being used by `IterableDataset` or `IterableDataset.sample` being explicitly passed.

Comment (cpuhrsch) [reply torch/utils/data/distributed.py]: Ok makes sense. Then set_rank should also work for https://github.com/pytorch/pytorch/issues/28743 (I saw you already commented). Maybe that's a bit beyond scope, but if we implement another sampler we could see if set_rank is a general, orthogonal extension we need for iterator-style samplers. For example some simple  sampling algorithm such as a variation of https://en.wikipedia.org/wiki/Reservoir_sampling

Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: `set_rank` can be used for reservoir sampling when reading duplicate examples in different workers are not desired. Some changes on the original algorithm might be necessary to make reservoir sampling fully distributed among workers, but that is a bit out of scope indeed :)

Using reservoir sampling could the next use case for https://github.com/pytorch/pytorch/pull/26547 once it is merged, as all the boiler plate code to distribute training with embedded sampler on iterable datasets will already be there. The original PR handles known chunk size, but if implement a reservoir sampler, `ChunkDataset` can leverage it to train on unlimited number of chunks

**Comment by thiagocrepaldi** [comment]
ping @fmassa :)

**Review by fmassa — CHANGES_REQUESTED** [review]
Hi,

Sorry for the delay in reviewing.

I don't think that `DistributedSampler` should necessarily work with `IterableDataset`. The indices returned by the `DistributedSampler` are by definition useless to the general `IterableDataset`, and trying to use it to fit a particular use-case would add unnecessary constraints to `IterableDataset`.

My understanding is that this PR is trying to accomplish two things:
1. add some meta-information to the `DistributedSampler` to work on `IterableDatasets`
2. (potentially) simplify the use of `IterableDatasets` in distributed mode (not present in this PR).

The concept of a sampler is not valid for `IterableDataset` in general -- we have no guarantees on the order of the examples that will be returned. I believe we should keep this as is in general.
Does this mean that no `IterableDataset` can know ahead of time of its iteration order? No, they can, but this is a special case, and should be handled by the application. I don't think this should live in PyTorch.

But then, how to make it easier for users to write their own `IterableDataset` that works on distributed?

What are the things we need to keep in mind in DDP in this case?
- `sampler.set_epoch`, so that we can split the dataset between different machines
- in particular, we need to handle the `idx` of the worker ourselves.

How can we accomplish both, without having to change the APIs of `DistributedSampler` nor `DataLoader`?

Here is one example.

```python
class MyIterableDataset(IterableDataset):
    def __init__(self):
        # each dataset, when constructed, know its rank
        # and they are constructed once per (GPU) process
        self.rank = dist.get_rank()
        # how many are we?
        self.world_size = dist.get_world_size()
        # have a counter on how many epochs have passed
        # can be incremented by get_chunk_iterator
        # if the user wants different shuffle per epochs
        self.epoch = 0

    def __iter__(self):
        worker_info = torch.utils.data.get_worker_info()
        chunk_id = self.rank
        total_chunks = self.world_size
        if worker_info is not None:
            chunk_id = self.rank * worker_info.num_workers + worker_info.id
            total_chunks = self.world_size * worker_info.num_workers

        # now return the chunks in the dataset accordingly
        return iter(self.get_chunk_iterator(chunk_id, total_chunks))

    def get_chunk_iterator(self, chunk_id, total_chunks):
        # user implements this
        pass
```
Then, all the logic specific to your application stays restricted to your Dataset implementation, via the `get_chunk_iterator`, which can handle buffering / shuffling / etc, and can also increment the epoch counter if the user wants.

I might be missing something, but let me know if the above implementation doesn't address your use-cases.

**Comment by cpuhrsch** [comment]
@fmassa - samplers are definitely relevant to IterableDatasets - there are various sampling techniques that apply to streams such as https://en.wikipedia.org/wiki/Reservoir_sampling

**Comment by fmassa** [comment]
@cpuhrsch I believe this should be an implementation detail of the specialization of the `IterableDataset` , this doesn't mean that the concept of sampler that we have should support `IterableDataset`.

**Comment by thiagocrepaldi** [comment]
> Hi,
> 
> Sorry for the delay in reviewing.
> 
> I don't think that `DistributedSampler` should necessarily work with `IterableDataset`. The indices returned by the `DistributedSampler` are by definition useless to the general `IterableDataset`, and trying to use it to fit a particular use-case would add unnecessary constraints to `IterableDataset`.
> 
> My understanding is that this PR is trying to accomplish two things:
> 
>     1. add some meta-information to the `DistributedSampler` to work on `IterableDatasets`
> 
>     2. (potentially) simplify the use of `IterableDatasets` in distributed mode (not present in this PR).
> 
> 
> The concept of a sampler is not valid for `IterableDataset` in general -- we have no guarantees on the order of the examples that will be returned. I believe we should keep this as is in general.
> Does this mean that no `IterableDataset` can know ahead of time of its iteration order? No, they can, but this is a special case, and should be handled by the application. I don't think this should live in PyTorch.
> 
> But then, how to make it easier for users to write their own `IterableDataset` that works on distributed?
> 
> What are the things we need to keep in mind in DDP in this case?
> 
>     * `sampler.set_epoch`, so that we can split the dataset between different machines
> 
>     * in particular, we need to handle the `idx` of the worker ourselves.
> 
> 
> How can we accomplish both, without having to change the APIs of `DistributedSampler` nor `DataLoader`?
> 
> Here is one example.
> 
> ```python
> class MyIterableDataset(IterableDataset):
>     def __init__(self):
>         # each dataset, when constructed, know its rank
>         # and they are constructed once per (GPU) process
>         self.rank = dist.get_rank()
>         # how many are we?
>         self.world_size = dist.get_world_size()
>         # have a counter on how many epochs have passed
>         # can be incremented by get_chunk_iterator
>         # if the user wants different shuffle per epochs
>         self.epoch = 0
> 
>     def __iter__(self):
>         worker_info = torch.utils.data.get_worker_info()
>         chunk_id = self.rank
>         total_chunks = self.world_size
>         if worker_info is not None:
>             chunk_id = self.rank * worker_info.num_workers + worker_info.id
>             total_chunks = self.world_size * worker_info.num_workers
> 
>         # now return the chunks in the dataset accordingly
>         return iter(self.get_chunk_iterator(chunk_id, total_chunks))
> 
>     def get_chunk_iterator(self, chunk_id, total_chunks):
>         # user implements this
>         pass
> ```
> 
> Then, all the logic specific to your application stays restricted to your Dataset implementation, via the `get_chunk_iterator`, which can handle buffering / shuffling / etc, and can also increment the epoch counter if the user wants.
> 
> I might be missing something, but let me know if the above implementation doesn't address your use-cases.

Thanks for the comment, Francisco!

In distributed training, the `dataset` is instantiated in the main process and is cloned for each worker, so `torch.utils.data.get_worker_info()` is not available during `MyIterableDataset` definition. The user would need a `worker_fn_init` implementation that has most of `__init__` and `__iter__` code. This pattern would be repeated for all IterableDataset experiment with distributed training. This PR will enable future `IterableDataset` implementations to follow the same pattern as regular Dataset without unnecessary boilerplate code.

As @cpuhrsch mentioned, there are several use cases for sampling within `IterableDataset`, including but not limited to Reservoir sampling that was requested https://github.com/pytorch/pytorch/issues/28743. Reservoir sampler could be implemented from the new `DistributedSampler` and it would be interesting as it allows sampling dataset with unknown size without duplicates in an efficient way (that by sampler might be a separate PR - but lets drop that idea for now :))

**Review by fmassa — APPROVED** [review]
We can move on with this PR, although I would have loved a review from either @apaszke or @SsnL

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Let's wait on @apaszke or @SsnL

File: torch/utils/data/distributed.py
Comment (SsnL) [inline torch/utils/data/distributed.py]: `IterableDataset` can still implement `__len__`. I think it makes more sense for this sampler to assume that the dataset has `__len__` than having an explicit input argument.

**Review by ssnl — COMMENTED** [review]
Thanks for the PR!

After reading through the thread, I am still a bit confused how this helps distributed sampler for `IterableDataset`s. Is this supposed to directly work already, or will there be follow-up patches?

The current PyTorch `Sampler` is different from the "sampling" in methods like reservoir sampling, in the sense that it only specifies the "indices" to sampler, rather than whether a sample should be kept. Due to this, in DataLoader, we currently disallow using `IterableDataset` with  a custom `sampler` or `batch_sampler`. An example snippet would be really useful!

Finally, this needs a test to get in.

File: torch/utils/data/distributed.py
Comment (thiagocrepaldi) [reply torch/utils/data/distributed.py]: @SsnL if the dataset can implement `len` then it makes more sense to extend `Dataset` as opposed to `IterableDataset`

This PR tries to add the ability to do *distributed* training when the number of samples are *unknown*. `IterableDataset` allows this concept, but for distributed training, the sampler needs some hints on how many chunks this unknown number of samples are split at

**Comment by thiagocrepaldi** [comment]
Closing this for now

**Comment by rabeehkarimimahabadi** [comment]
Hi
I have multiple large-scale datasets and I know their length in advance, I need to train the models on TPUs and need to write distributed sampler for this case, to be able to use pytorch XLA, I really appreciate providing me with some examples, how I can implement distributed sampler for iterative datasets.
In the non-iterative case this is implemented as follows 

def get_tpu_sampler(dataset: torch.utils.data.dataset.Dataset):
    if xm.xrt_world_size() <= 1:
        return RandomSampler(dataset)
    return DistributedSampler(dataset, num_replicas=xm.xrt_world_size(), rank=xm.get_ordinal())

could you assist me how I can implement it for my case?
thanks. I really appreciate some help on this.  
Best
Rabeeh

**Pull request closed without merging**
